examples/shader/shader_material_manual.rs

```diff
@@ -0,0 +1,187 @@
+        },
+        renderer::RenderDevice,
+        texture::FallbackImage,
+    },
+};
+use bevy_internal::render::render_resource::{
+    BindGroupLayoutDescriptor, BindGroupLayoutEntry, BindingType, BufferBindingType,
+    SamplerBindingType, ShaderStages, ShaderType, TextureSampleType, TextureViewDimension,
+};
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_plugin(MaterialPlugin::<CustomMaterial>::default())
+        .add_startup_system(setup)
+        .run();
+}
+
+/// set up a simple 3D scene
+fn setup(
+    mut commands: Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<CustomMaterial>>,
+    asset_server: Res<AssetServer>,
+) {
+    // cube
+    commands.spawn().insert_bundle(MaterialMeshBundle {
+        mesh: meshes.add(Mesh::from(shape::Cube { size: 1.0 })),
+        transform: Transform::from_xyz(0.0, 0.5, 0.0),
+        material: materials.add(CustomMaterial {
+            color: Color::BLUE,
+            color_texture: Some(asset_server.load("branding/icon.png")),
+            alpha_mode: AlphaMode::Blend,
+        }),
+        ..default()
+    });
+
+    // camera
+    commands.spawn_bundle(Camera3dBundle {
+        transform: Transform::from_xyz(-2.0, 2.5, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
+        ..default()
+    });
+}
+
+/// The Material trait is very configurable, but comes with sensible defaults for all methods.
+/// You only need to implement functions for features that need non-default behavior. See the Material api docs for details!
+impl Material for CustomMaterial {
+    fn fragment_shader() -> ShaderRef {
+        "shaders/custom_material.wgsl".into()
+    }
+
+    fn alpha_mode(&self) -> AlphaMode {
+        self.alpha_mode
+    }
+}
+
+/// This is the struct that will be passed to your shader
+/// Specific order of declaration is not relevant, but "binding" identifier values are.
+/// Defined in `AsBindGroup` implementation, they should match their corresponding shader @binding value.
+#[derive(Debug, Clone, TypeUuid)]
+#[uuid = "f690fdae-d598-45ab-8225-97e2a3f056e0"]
+struct CustomMaterial {
+    color: Color,
+    color_texture: Option<Handle<Image>>,
+    alpha_mode: AlphaMode,
+}
+
+impl AsBindGroup for CustomMaterial {
+    type Data = ();
+    fn as_bind_group(
+        &self,
+        layout: &BindGroupLayout,
+        render_device: &RenderDevice,
+        images: &RenderAssets<Image>,
+        fallback_image: &FallbackImage,
+    ) -> Result<PreparedBindGroup<Self>, AsBindGroupError> {
+        // Step 1: retrieve information from our extracted type.
+        let color_texture = {
+            let color_texture_handle: Option<&Handle<Image>> = (&self.color_texture).into();
+            if let Some(handle) = color_texture_handle {
+                images
+                    .get(handle)
+                    .ok_or(AsBindGroupError::RetryNextUpdate)?
+            } else {
+                fallback_image
+            }
+        };
```

Comment:
to note: this logic was duplicated with cargo-expand, to get :
- the sampler
- the texture

I refactored a bit to make only one call to retrieve the image and have 1 failure path.